Issue #6379: AnnotationLocation: add support for PACKAGE_DEF, ENUM_CO…

pbludov opened the pull request:
…NSTANT_DEF

Issue #6379 

regression tests https://pbludov.github.io/issue-6379/

rnveach commented:
Restarting travis had no effect.

pbludov commented:
> Restarting travis had no effect.

https://bugs.openjdk.java.net/browse/JDK-8195736
It looks like Travis pipeline should be upgraded to less ancient version of Ubuntu

romani commented:
TC is restarted

rnveach commented:
@pbludov travis is working now.

rnveach commented on src/main/resources/google_checks.xml:
```diff
@@ -210,8 +210,8 @@
-            <property name="tokens"
-             value="CLASS_DEF, INTERFACE_DEF, ENUM_DEF, METHOD_DEF, CTOR_DEF"/>
+            <property name="tokens" value="CLASS_DEF, CTOR_DEF, ENUM_CONSTANT_DEF, ENUM_DEF"/>
+            <property name="tokens" value="INTERFACE_DEF, METHOD_DEF, PACKAGE_DEF"/>
```
Since you are changing google_checks, please clarify which item in the google style guide says new tokens are ok.

rnveach commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/annotation/AnnotationLocationCheck.java:
```diff
@@ -239,6 +240,8 @@ public final void setAllowSamelineMultipleAnnotations(boolean allow) {
             TokenTypes.CLASS_DEF,
             TokenTypes.INTERFACE_DEF,
+            TokenTypes.PACKAGE_DEF,
+            TokenTypes.ENUM_CONSTANT_DEF,
```
@pbludov @romani Any reasons why these shouldn't be default?

pbludov replied:
`ENUM_CONSTANT_DEF` definitely should be.
For `PACKAGE_DEF` I think it is better to add.

romani commented:
@pbludov , thanks a lot for update.
But now smth is wrong with defition of what is planned to do issue.
Title is  declaring adding tokens but fix, do extra removal of token. We try to accurate with issues as they are part of release notes, and contain instructions on how to migrate if breaking changes happening. Removal of token is breaking changes.

Please move removal of this token to separate issue, as we need to update documentation to clearly define targets of this Check.

Please recheck `InputAnnotationLocationPackageAnnotations` I did not find usage of this new input.

pbludov commented:
> Please recheck `InputAnnotationLocationPackageAnnotations` I did not find usage of this new input.

Package annotations used in `package-info.java`.
https://github.com/checkstyle/checkstyle/pull/6380/files#diff-c6463b962b289dda35483cf9773d342d

pbludov commented:
> Please move removal of this token to separate issue, as we need to update documentation to clearly define targets of this Check.

Done.

romani commented on src/test/resources/com/puppycrawl/tools/checkstyle/checks/annotation/annotationlocation/InputAnnotationLocationPackageAnnotations.java:
```diff
@@ -0,0 +1,17 @@
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Target;
+
+class InputAnnotationLocationPackageAnnotations {
```
please make a comment that this file is not an input, it is just place to define annotations to be used in package-info.java. Eventually we end up when config will be defined in each Input file,  but looks like we have exceptions from this rule for files like this .

rnveach replied:
@romani @pbludov Secondary inputs can be placed in a special folder that will be skipped from validation and allowed to be named anything you want.
See https://github.com/checkstyle/checkstyle/blob/a73ff0890c27cbe8affc1a55f2710a231687bd85/src/test/java/com/puppycrawl/tools/checkstyle/internal/AllTestsTest.java#L258-L259 .

@romani Is this acceptable and you want a specific folder name? Other areas have used `inputs`.

romani replied:
@pbludov , please try to move to subfolder "inputs"
example  - https://github.com/checkstyle/checkstyle/blob/master/src/test/resources/com/puppycrawl/tools/checkstyle/checks/design/visibilitymodifier/inputs/InetSocketAddress.java

pbludov replied:
done

romani commented on src/test/resources/com/puppycrawl/tools/checkstyle/checks/annotation/annotationlocation/InputAnnotationLocationEnumField.java:
```diff
@@ -0,0 +1,30 @@
+ * tokens = ["ENUM_CONSTANT_DEF", "ENUM_DEF"]
+ *
+ */
+enum InputAnnotationLocationEnumField {
```
please add 2 annotations on enum also  (with and without violation)

pbludov replied:
done

romani commented on src/test/java/com/puppycrawl/tools/checkstyle/checks/annotation/AnnotationLocationCheckTest.java:
```diff
@@ -86,7 +86,7 @@ public void testIncorrectAllTokens() throws Exception {
                 + "CTOR_DEF, VARIABLE_DEF, PARAMETER_DEF, ANNOTATION_DEF, TYPECAST, "
                 + "LITERAL_THROWS, IMPLEMENTS_CLAUSE, TYPE_ARGUMENT, LITERAL_NEW, DOT, "
-                + "ANNOTATION_FIELD_DEF");
+                + "ANNOTATION_FIELD_DEF, ENUM_CONSTANT_DEF");
```
test is named `testIncorrectAllTokens ` probably we need to add all tokens.
and we probably need to update `InputAnnotationLocationIncorrect.java` with some cases.

rnveach replied:
Yes I added this test for pitest. All tokens should be added.

pbludov replied:
done.
Actually, `PACKAGE_DEF` is irrelevant here. All package annotations should be in the file `package-info.java`.

romani commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/annotation/AnnotationLocationCheck.java:
```diff
@@ -262,33 +266,35 @@ public final void setAllowSamelineMultipleAnnotations(boolean allow) {
-
-        if (modifiersNode != null) {
-            checkAnnotations(modifiersNode, getExpectedAnnotationIndentation(modifiersNode));
+        DetailAST annotationsNode = ast.findFirstToken(TokenTypes.MODIFIERS);
```
not sure this is good update to rename variable.
we searched for modifiers, and we should found modifiers node.
it is not good to reuse variable.

pbludov replied:
done

romani commented:
@pbludov , ping, let me know if smth is not clear.

rnveach commented:
@pbludov Last regression still had typecast in it. Can you re-run it since we removed it in this PR?

rnveach commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/annotation/AnnotationLocationCheck.java:
```diff
@@ -263,32 +267,37 @@ public final void setAllowSamelineMultipleAnnotations(boolean allow) {
      * The expected indentation should be the same as the indentation of the node
      * which is the parent of the target modifier node.
-     * @param modifierNode modifier node.
+     * @param annotationsNode modifiers or annotations node.
```
This change doesn't make sense. It is modifiers or annotation, yet previous name or new name just picks 1 side.
Can we just rename this to `node`?

rnveach commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/annotation/AnnotationLocationCheck.java:
```diff
@@ -263,32 +267,37 @@ public final void setAllowSamelineMultipleAnnotations(boolean allow) {
     /**
      * Returns an expected annotation indentation.
      * The expected indentation should be the same as the indentation of the node
      * which is the parent of the target modifier node.
```
`modifier node` => `node`

pbludov replied:
done

rnveach replied:
@pbludov This is not done.

pbludov replied:
my bad, now really done.

rnveach commented on src/test/java/com/puppycrawl/tools/checkstyle/internal/AllChecksTest.java:
```diff
@@ -175,7 +175,8 @@
                 "TYPECAST", "DOT", "TYPE_ARGUMENT", "ANNOTATION_DEF", "LITERAL_NEW",
-                "LITERAL_THROWS", "PARAMETER_DEF", "IMPLEMENTS_CLAUSE", "ANNOTATION_FIELD_DEF")
+                "LITERAL_THROWS", "PARAMETER_DEF", "IMPLEMENTS_CLAUSE", "ANNOTATION_FIELD_DEF",
+                "ENUM_CONSTANT_DEF", "PACKAGE_DEF")
```
The reasons for the new tokens should be the same as https://github.com/checkstyle/checkstyle/pull/6380#discussion_r248938124

rnveach replied:
@romani This check is in google config and we added them as defaults. Will this cause an issue like in https://github.com/checkstyle/checkstyle/issues/6381#issuecomment-455003077 ?

romani replied:
yes, your are right, .... amount of items to rememeber to recheck is still high.
@pbludov , please recheck that this does not conflict with google style guide.

pbludov replied:
The google style is very ambiguous:

> Annotations applying to a class, method or constructor appear immediately after the documentation block, and each annotation is listed on a line of its own (that is, one annotation per line). 
Exception: A single parameterless annotation may instead appear together with the first line of the signature
Annotations applying to a field also appear immediately after the documentation block, but in this case, multiple annotations (possibly parametrized) may be listed on the same line
There are no specific rules for formatting annotations on parameters, local variables, or types. 

That's all. Even the style for the interface definition is unclear.

rnveach replied:
If there are no rules pertaining to package then the token can't be added to the required. Default is ok as long as google config doesn't use the default list, which it doesn't.

For ENUM_CONSTANT_DEF, should we treat this as a field under the rules? To me it makes sense. @romani Do you agree?

> The google style is very ambiguous:

Yep. :)

romani replied:
In latest code I see new tokens in acceptable and default it is correct. Package token will not be considered.

For Google style we need to have proper config, so we can change config and all Check functionality as we need, but just need to got required functionality. If Check  is used default in  config, and we add new default token, we can change config to exclude new token.

romani replied:
>For ENUM_CONSTANT_DEF, should we treat this as a field under the rules?

No, it is not a field. But we might ask confirmation at their issue tracker.

rnveach replied:
@pbludov For this item, just add the reasons for not adding the tokens to the google config in the test.

pbludov commented:
> @pbludov Last regression still had typecast in it. Can you re-run it since we removed it in this PR?

Actually, it is still here. I reverted the removal, as @romani suggested:
https://github.com/checkstyle/checkstyle/pull/6380#issuecomment-458371616

rnveach commented:
Yes token was added back in, but didn't your original regression have it removed?

Original commit in regression had `"Issue #6379: AnnotationLocation: add support for PACKAGE_DEF, ENUM_CONSTANT_DEF, TYPECAST"` but you removed typecast from the commit message when romani asked you to undo the removal and put it in another PR.

pbludov commented:
@rnveach Please check again https://pbludov.github.io/issue-6379/

rnveach commented:
> http://checkstyle.sourceforge.net/config_annotation.html#AnnotationLocation
>  By default, Check enforce to locate ... annotation should be located on separate line from target element. 

https://pbludov.github.io/issue-6379/hibernate-orm/index.html#A2
Why isn't this a `should be alone on line` violation?

rnveach commented:
https://pbludov.github.io/issue-6379/openjdk8/index.html#A3
This isn't an enum or package, so why is it showing up?

pbludov commented:
> https://pbludov.github.io/issue-6379/openjdk8/index.html#A3
> This isn't an enum or package, so why is it showing up?

The question should be "why wasn't it shown before?" The token `IMPLEMENTS_CLAUSE` is in the acceptable tokens for years. It comes from https://github.com/checkstyle/checkstyle/pull/604/files#diff-e3369b5d06b91380b99268196514e720R185

rnveach commented:
@pbludov There is no IMPLEMENTS_CLAUSE there, only a TYPE_UPPER_BOUNDS.
https://pbludov.github.io/issue-6379/openjdk8/index.html#A3
https://pbludov.github.io/issue-6379/openjdk8/xref/home/pbludov/src/github.com/pbludov/contribution/checkstyle-tester/repositories/openjdk8/test/java/lang/annotation/TypeAnnotationReflection.java.html#L516

520 has an implements, which is also producing a new violation. Why are these violations only showing up with your changes for different tokens?
If they should be showing up, then it looks like a bug they were missed and somehow this may have to be split into another issue.
@romani Please review.

romani commented:
I will review this PR again.

romani commented:
in diff report I see difference in tokens as +2 tokens - `PACKAGE_DEF,ENUM_CONSTANT_DEF`.
Which is expected

Not clear why report is so interesting.

> please recheck that this does not conflict with google style guide.

This Check is used in Google Style, we need to accurate with update -https://checkstyle.org/google_style.html#a4.8.5 , especially as such tokens are become default. BUT fortunately for us we do not use default config - https://github.com/checkstyle/checkstyle/blob/3835575706d8240b3ec054d17ed48f692b168611/src/main/resources/google_checks.xml#L211-L220 so we have some freedom on regression.

romani commented:
@pbludov , do you need some help from us ? 

Sorry that such simple update (at the beginning) result in such a mess.
If we have move issues in this Check, it is ok we can address them separately, but we should avoid false-positives as much as possible, as users usually disable Check for misbehavior (and of  cause they forgot to activate it back), so Check become actually dead.

rnveach commented:
@pbludov please rebase to fix conflicts.

pbludov commented:
> @pbludov please rebase to fix conflicts.

done. Updated regression report:
https://pbludov.github.io/issue-6379-v2/

pbludov commented:
> This isn't an enum or package, so why is it showing up?

Solved. Please review again.

rnveach commented:
https://github.com/checkstyle/checkstyle/pull/6380#issuecomment-460069159
> https://pbludov.github.io/issue-6379/hibernate-orm/index.html#A2
Why isn't this a `should be alone on line` violation?
https://pbludov.github.io/issue-6379/hibernate-orm/index.html#A4
This one especially.

These violations haven't changed and @romani agreed with this assessment with a thumbs up.

romani commented:
>`should be alone on line`

Is preferable, but existing violation is also ok as form of better than nothing.

pbludov commented:
> Why isn't this a `should be alone on line` violation?

This is because the property `allowSamelineSingleParameterlessAnnotation` set to `true` by default.
And  `@NamedQuery(name = "allCarpet", query = "select c from Carpet c")` is a single *parameterless* annotation. No kidding. I'll start a new issue for this.

rnveach commented:
> https://pbludov.github.io/issue-6379/hibernate-orm/index.html#A4
> This is because the property `allowSamelineSingleParameterlessAnnotation` set to `true` by default.

If `allowSamelineSingleParameterlessAnnotation` is set to true by default and this is a single parameterless annotation then there should be no violation on the package the same as there is no violation on a method.
By this logic there shouldn't be any violation period, same as the following example on a method.

````
$ cat TestClass.java
import org.hibernate.annotations.NamedQuery;

public class TestClass {
  @NamedQuery(name = "allCarpet", query = "select c from Carpet c") public void method() {
    }
}

$ cat TestConfig.xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
          "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">

<module name="Checker">
    <property name="charset" value="UTF-8"/>

    <module name="TreeWalker">
<module name="AnnotationLocation">
  <property name="id" value="default" />
</module>
<module name="AnnotationLocation">
  <property name="id" value="disallowSamelineSingleParameterlessAnnotation" />
  <property name="allowSamelineSingleParameterlessAnnotation" value="false" />
</module>
    </module>
</module>

$ java -jar checkstyle-8.17-all.jar -c TestConfig.xml TestClass.java
Starting audit...
[ERROR] TestClass.java:4: Annotation 'NamedQuery' should be alone on line. [disallowSamelineSingleParameterlessAnnotation]
Audit done.
Checkstyle ends with 1 errors.
````

pbludov commented:
@rnveach you are right:
```
@Ann package test; // false violation
@Ann class Cls1 {} // ok
@interface Ann {}
```
Note the ast:
```
PACKAGE_DEF -> package [1:5]                             // PACKAGE_DEF has column = 5
|--ANNOTATIONS -> ANNOTATIONS [1:0]
|   `--ANNOTATION -> ANNOTATION [1:0]
|       |--AT -> @ [1:0]
|       `--IDENT -> Ann [1:1]
|--IDENT -> test [1:13]
`--SEMI -> ; [1:17]
CLASS_DEF -> CLASS_DEF [2:0]                             // CLASS_DEF has column = 0
|--MODIFIERS -> MODIFIERS [2:0]
|   `--ANNOTATION -> ANNOTATION [2:0]
|       |--SINGLE_LINE_COMMENT -> // [1:19]
|       |   `--COMMENT_CONTENT ->  false violation\n [1:21]
|       |--AT -> @ [2:0]
|       `--IDENT -> Ann [2:1]
|--LITERAL_CLASS -> class [2:5]
|--IDENT -> Cls1 [2:11]
`--OBJBLOCK -> OBJBLOCK [2:16]
    |--LCURLY -> { [2:16]
    `--RCURLY -> } [2:17]
```
For some reason, the column for `PACKAGE_DEF` is `5` instead of `0`.

rnveach commented:
> For some reason, the column for `PACKAGE_DEF` is `5` instead of `0`.

The devil is in the details.
`CLASS_DEF` has no text from the source file, so it is a completely generated token from the antlr. As such, it generates its line/column from the first sub-token it finds in it's tree.
`PACKAGE_DEF` has text from the source file (`package`), so it's line/column comes the source file and is not generated.

Unless tree is changed, check will have to make this type of distinction between package and class.
Can you update the check?

pbludov commented:
> Unless tree is changed, check will have to make this type of distinction between package and class.
> Can you update the check?

Yes, the check should be updated.

@rnveach @romani take a look at this snippet:
```
package test;
class Cls1 {}
```
The AST is
```
PACKAGE_DEF -> package [1:0]
|--ANNOTATIONS -> ANNOTATIONS [1:8]
|--IDENT -> test [1:8]
`--SEMI -> ; [1:12]
CLASS_DEF -> CLASS_DEF [2:0]
|--MODIFIERS -> MODIFIERS [2:0]
|--LITERAL_CLASS -> class [2:0]
|--IDENT -> Cls1 [2:6]
`--OBJBLOCK -> OBJBLOCK [2:11]
    |--LCURLY -> { [2:11]
    `--RCURLY -> } [2:12]
```
`PACKAGE_DEF` has empty `ANNOTATIONS` block between "package" and "test".
`package` column is 0, `test` column is 8, empty annotations block column is also 8.
 This is incorrect and may cause errors.

pbludov commented:
For this case
```
  @Annotation1
@Annotation2 class InputAnnotationLocationClass {}
```
The violation is `Annotation 'Annotation2' have incorrect indentation level 0, expected level should be 2.`
This is definitely wrong. Indentation level of `Annotation2` is right.
I think the rule should be "All annotations should be aligned on the left side"

romani commented:
I am for new issue.

rnveach commented:
@pbludov new issues are always ok and we can discuss them in more detail there. Is there any chance we can finish this PR? :)

pbludov commented:
@romani @rnveach done.
new regression report: https://pbludov.github.io/issue-6379-3/

There is still a problem with annotations after modifiers:
```
final @Ann Object field; // expected violation here
```
This is due to
https://github.com/checkstyle/checkstyle/blob/master/src/main/java/com/puppycrawl/tools/checkstyle/checks/annotation/AnnotationLocationCheck.java#L285
```
        DetailAST annotation = modifierNode.getFirstChild();

        while (annotation != null && annotation.getType() == TokenTypes.ANNOTATION) {
        ....
        }
```
when the first child of a `MODIFIERS` node is not an annotation, the check does nothing (this feature is documented). I'll start a new issue for this.

rnveach commented:
New regression looks good, but I am surprised all is `should be alone on line` and no indentation errors.

rnveach commented:
@romani Please re-review since we made changes to the logic.
